Task migration from using Resources to using workspaces

Preeticp opened the pull request:
This PR fixes: 

 - https://issues.redhat.com/browse/RHDEVDOCS-2081
 - https://issues.redhat.com/browse/RHDEVDOCS-1693
 - https://issues.redhat.com/browse/RHDEVDOCS-2109

 The YAML files in the creating workflow and Understanding conceptual docs have been updated to reflect the migration from Resources to Workspaces. Notes for migration added. 
New module for pvc volumes added.
This also fixes the bug of adding cloning the tutorial repo to the Prerequisites in the Creating flow.

- This has been reviewed by the QE 
- This needs to be merged to 4.5 and 4.6

Preeticp commented:
@ppitonak @khrm @piyush-garg request a review please.

ppitonak reviewed (changes requested):
In section Creating Webhooks, we instruct user to create a webhook in frontend repo `Configure Webhooks manually on the front-end repository:` but later we commit to api repo.

ppitonak commented on modules/op-running-a-pipeline.adoc:
```diff
@@ -15,7 +13,7 @@ A PipelineRun starts a Pipeline and ties it to the Git and image resources that
 +
 ----
-$ tkn pipeline start build-and-deploy -r git-repo=api-repo -r image=api-image -p deployment-name=vote-api
+$ tkn pipeline start build-and-deploy -w name=shared-workspace,claimName=source-pvc -p deployment-name=vote-api -p git-url=http://github.com/openshift-pipelines/vote-api.git -p IMAGE=image-registry.openshift-image-registry.svc:5000/pipelines-tutorial/vote-api
```
example https://github.com/openshift/pipelines-tutorial/blob/release-tech-preview-2/01_pipeline/03_persistent_volume_claim.yaml

Preeticp replied:
Created another section

khrm commented on modules/op-assembling-a-pipeline.adoc:
```diff
@@ -5,7 +5,9 @@
 
-A Pipeline represents a CI/CD flow and is defined by the Tasks to be executed. It specifies how the Tasks interact with each other and their order of execution using the `inputs` , `outputs`, and `runAfter` parameters. It is designed to be generic and reusable in multiple applications and environments.
+A Pipeline represents a CI/CD flow and is defined by the Tasks to be executed. It is designed to be generic and reusable in multiple applications and environments.
+It specifies how the Tasks interact with each other and their order of execution using the `inputs` , `outputs`, and `runAfter` parameters.
```
Yes, they referred to Pipelineresources. If you want to define the order for input and output, we need `from` clause.

Preeticp commented on pipelines/creating-applications-with-cicd-pipelines.adoc:
```diff
@@ -10,35 +10,30 @@ With {pipelines-title}, you can create a customized CI/CD solution to build, tes
 
-[discrete]
-== Prerequisites
+= Prerequisites
```
I had a conflict and looks like that's when this was introduced!

Preeticp commented on pipelines/creating-applications-with-cicd-pipelines.adoc:
```diff
@@ -10,35 +10,30 @@ With {pipelines-title}, you can create a customized CI/CD solution to build, tes
 include::modules/op-creating-pipeline-tasks.adoc[leveloffset=+1]
 
-include::modules/op-defining-and-creating-pipelineresources.adoc[leveloffset=+1]
+//include::modules/op-defining-and-creating-pipelineresources.adoc[leveloffset=+1]
```
Will do

Preeticp replied:
Since it was not yet deprecated, had commented it out.

ppitonak commented on modules/op-about-triggers.adoc:
```diff
@@ -95,9 +103,9 @@ spec:
   triggers:
   - bindings: <5>
-    - name: vote-app-binding
+    - ref: vote-app
```
this was marked as resolved but it didn't change

Preeticp commented:
Good catch fixed the commit section in Triggering a PipelineRun

Preeticp commented:
@ppitonak I have added a section for the PVC volumes and added notes wrt migration from resources to workspaces in 2 places as discussed. Request you to review please.

ppitonak commented on modules/op-specifying-pvc-as-volumesource-in-workspaces.adoc:
```diff
@@ -0,0 +1,41 @@
+Alternatively, you can also execute the YAML file directly from the Git repository:
++
+----
+$ oc create -f https://raw.githubusercontent.com/openshift/pipelines-tutorial/master/01_pipeline/03_persistent_volume_claim.yaml
```
This needs to be updated to tech-preview branch... https://raw.githubusercontent.com/openshift/pipelines-tutorial/release-tech-preview-2/01_pipeline/03_persistent_volume_claim.yaml

ppitonak commented:
LGTM

bobfuru commented on modules/op-assembling-a-pipeline.adoc:
```diff
@@ -33,54 +36,78 @@ kind: Pipeline
 ----
 +
-Notice that the Pipeline definition abstracts away the specifics of the Git source repository and image registries to be used during the Pipeline execution.
+Notice that the Pipeline definition abstracts away the specifics of the Git source repository and image registries to be used during the Pipeline execution as `params`.
+The details of the git repositories and image registries to be used are added when a pipeline is triggered during pipeline execution.
```
- If you are calling attention to this chunk, it should be wrapped in a note/admonition style.
- The language is unclear and I'm not sure what I should be noticing. Suggest switching to more active voice, something like, "The Pipeline definition abstracts away the details of the Git source repository and image registries. These details are added as `params` when execution triggers a Pipeline."

bobfuru commented on modules/op-release-notes-1-1.adoc:
```diff
@@ -17,6 +17,8 @@
 
 === Pipelines
+
+* Support for using Workspaces instead of PipelineResources is now provided. We recommend using Workspaces in OpenShift Pipelines, as PipelineResources are opaque, difficult to debug, limited in scope, and make Tasks less reusable. For more details on Workspaces, see the detailed OpenShift Pipelines concepts section.
```
A suggested revision for readability and style:

> Workspaces can now be used instead of PipelineResources. Use of Workspaces is recommended in OpenShift Pipelines, as PipelineResources are difficult to debug, limited in scope, and can make Tasks less reusable. For more details on Workspaces, see **Understanding OpenShift Pipelines** (with link to assembly).

(For the last part, I don't think the reader will be able to understand what is meant by "concepts section.")

Preeticp replied:
I would have loved to add a link but xrefs can not be added to modules, right?

bobfuru replied:
Oh, right. Sorry, didn't realize this was a module so ignore my link suggestion.

bobfuru commented on modules/op-specifying-pvc-as-volumesource-in-workspaces.adoc:
```diff
@@ -0,0 +1,41 @@
+[id="specifying-pvc-as-volumesource-in-workspaces_{context}"]
+= Specifying PersistentVolumeClaims as VolumeSource in Workspaces
+
+Workspaces help share data among Tasks and allow you to specify one or more volumes that each Task in the Pipeline requires during execution.
```
s/Workspaces help share data among Tasks and /Workspaces help Tasks share data, and/

bobfuru commented on modules/op-specifying-pvc-as-volumesource-in-workspaces.adoc:
```diff
@@ -0,0 +1,41 @@
+
+Workspaces help share data among Tasks and allow you to specify one or more volumes that each Task in the Pipeline requires during execution.
+
+In this section, you will create a PersistentVolumeClaim to provide data storage and bind it to the workspace. This PersistentVolumeClaim provides the volumes or filesystem required for the pipeline execution.
```
I'm glad to see "workspace" and "pipeline" in lowercase here, though it doesn't follow the conventions of capitalizing objects elsewhere in the assembly. I suggest capitalizing here for consistency (at least until the business is convinced otherwise 😉 ).

bobfuru commented on modules/op-specifying-pvc-as-volumesource-in-workspaces.adoc:
```diff
@@ -0,0 +1,41 @@
+[discrete]
+.Procedure
+
+. Copy the contents of the following sample PersistentVolumeClaim YAML file and save it:
```
Suggest: "Copy and save the contents..."
Alternately, you might not want to instruct the user to copy this if it is sample data. So you might say something like, "Create a PersistentVolumeClaim YAML file similar to the following example:"

Preeticp replied:
For this entire workflow to work they need to use this particular YAML file hence instructions to copy. Otherwise I like the second option too.

bobfuru commented on modules/op-specifying-pvc-as-volumesource-in-workspaces.adoc:
```diff
@@ -0,0 +1,41 @@
+      storage: 500Mi
+----
++
+. Create the PersistentVolumeClaim:
```
For clarity:
s/. Create the PersistentVolumeClaim:/. Create the PersistentVolumeClaim, specifying the file you just created:/

bobfuru commented on modules/op-triggering-a-pipelinerun.adoc:
```diff
@@ -7,20 +7,20 @@
 Whenever a `push` event occurs in the Git repository, the configured Webhook sends an event payload to the publicly exposed EventListener service route. The EventListener service of the application processes the payload, and passes it to the relevant TriggerBindings and TriggerTemplates pair. The TriggerBinding extracts the parameters and the TriggerTemplate uses these parameters to create resour ...
 
-In this section, you will push an empty commit to the front-end `vote-api` repository, which will trigger the PipelineRun.
+In this section, you will push an empty commit to the front-end `vote-ui` repository, which will trigger the PipelineRun.
```
I know this assembly uses the convention of, "In this section, you will..." but it doesn't seem to align with our overall docs practice of using a more active voice (as opposed to "you will").

If you're so inclined, I'd suggest revising to something like, 

> The following example describes how to push an empty commit to the front-end `vote-ui` repository, which triggers the PipelineRun.

bobfuru commented on pipelines/creating-applications-with-cicd-pipelines.adoc:
```diff
@@ -10,38 +10,34 @@ With {pipelines-title}, you can create a customized CI/CD solution to build, tes
 * Create custom Tasks, or install existing reusable Tasks.
-* Create a Pipeline and PipelineResources to define the delivery Pipeline for your application.
+* Create and define the delivery Pipeline for your application.
+* Create a PersistentVolumeClaim attached to the Workspace to provide the volume or filesystem for pipeline execution.
```
Here pipeline is lowercased in "pipeline execution."

bobfuru commented on pipelines/creating-applications-with-cicd-pipelines.adoc:
```diff
@@ -10,38 +10,34 @@ With {pipelines-title}, you can create a customized CI/CD solution to build, tes
-* A front-end interface `vote-ui`, with the source code in link:https://github.com/openshift-pipelines/vote-ui/tree/release-tech-preview-1[`ui-repo`] Git repository.
-* A back-end interface `vote-api`, with the source code in link:https://github.com/openshift-pipelines/vote-api/tree/release-tech-preview-1[`api-repo`]  Git repository.
-* `apply_manifest` and `update-deployment` Tasks in link:https://github.com/openshift/pipelines-tutorial/tree/release-tech-preview-1[`pipelines-tutorial`] Git repository.
+* A front-end interface `vote-ui`, with the source code in the link:https://github.com/openshift-pipelines/vote-ui/tree/release-tech-preview-2[`ui-repo`] Git repository.
+* A back-end interface `vote-api`, with the source code in the link:https://github.com/openshift-pipelines/vote-api/tree/release-tech-preview-2[`api-repo`]  Git repository.
```
There should also be a comma after interface in these examples ("...front-end interface, `vote-ui`, with...").

bobfuru commented on pipelines/creating-applications-with-cicd-pipelines.adoc:
```diff
@@ -10,38 +10,34 @@ With {pipelines-title}, you can create a customized CI/CD solution to build, tes
-* You have forked the front-end link:https://github.com/openshift-pipelines/vote-ui/tree/release-tech-preview-1[`ui-repo`] and back-end link:https://github.com/openshift-pipelines/vote-api/tree/release-tech-preview-1[`api-repo`] Git repositories using your GitHub ID.
-
-* You have Administrator access to your repositories.
+* You have forked the front-end link:https://github.com/openshift-pipelines/vote-ui/tree/release-tech-preview-2[`ui-repo`] and back-end link:https://github.com/openshift-pipelines/vote-api/tree/release-tech-preview-2[`api-repo`] Git repositories using your GitHub ID and have Administrator access to these repositories.
```
s/GitHub ID and have/GitHub ID, and you have/

bobfuru commented on pipelines/creating-applications-with-cicd-pipelines.adoc:
```diff
@@ -51,7 +47,7 @@ include::modules/op-creating-webhooks.adoc[leveloffset=+1]
 
 [id="pipeline-addtl-resources"]
-.Additional resources
+== Additional resources
```
Thanks for updating this. 😁

Preeticp commented:
@bobfuru Thank you. 
I have fixed all the comments except the ones which deal with code snippets. I will discuss the code changes with the devs and create an issue for them, I think they might have to resolve it in the Upstream too. Build/built was something on my mind too.
I also have a question wrt adding links to modules. 
PTAL and merge if this looks alright to you.

bobfuru commented on modules/op-about-workspace.adoc:
```diff
@@ -23,7 +28,7 @@ You can specify Workspaces in the TaskRun or PipelineRun using:
 * An emptyDir that is discarded when the TaskRun completes
 
-Here is a code snippet of the `build-and-deploy` Pipeline which declares a `shared-workspace` Workspace for the Tasks, `build-image` and `apply-manifests`, defined in the Pipeline.
+The following example shows a code snippet of the `build-and-deploy` Pipeline which declares a `shared-workspace` Workspace for the Tasks, `build-image` and `apply-manifests`, defined in the Pipeline.
```
Commas in this sentence are a bit awkward; didn't catch it before. Suggestion:
The following example shows a code snippet of the `build-and-deploy` Pipeline, which declares a `shared-workspace` Workspace for the `build-image` and `apply-manifests` Tasks as defined in the Pipeline.

bobfuru commented on modules/op-about-workspace.adoc:
```diff
@@ -65,8 +70,8 @@ spec:
 <4> Name that uniquely identifies the Workspace used in the Task. This Task uses one Workspace named `source`.
-<5> Name of the Pipeline Workspace used by the Task. Note that the Workspace `source` in turn uses Pipeline Workspace `shared-workspace`.
-<6> List of Workspaces used in `apply-manifests` Task. Note that this Task shares the `source` Workspace with `build-image` Task.
+<5> Name of the Pipeline Workspace used by the Task. Note that the Workspace `source` in turn uses the Pipeline Workspace - `shared-workspace`.
```
I would avoid the `-` in this sentence and say something like, "...the Pipeline Workspace named `shared-workspace`.

bobfuru commented on modules/op-release-notes-1-1.adoc:
```diff
@@ -17,6 +17,8 @@
 
 === Pipelines
+
+* Workspaces can now be used instead of PipelineResources is now provided. It is recommended that you use Workspaces in OpenShift Pipelines, as PipelineResources are difficult to debug, limited in scope, and make Tasks less reusable. For more details on Workspaces, see Understanding OpenShift Pipelines.
```
For the first sentence, I think you can remove "is now provided."

Preeticp replied:
Thank you for pointing that!! Not sure how I missed that!

bobfuru commented on modules/op-specifying-pvc-as-volumesource-in-workspaces.adoc:
```diff
@@ -0,0 +1,41 @@
+      storage: 500Mi
+----
++
+. Create the PersistentVolumeClaim, specifying the file you just created::
```
Remove extra colon

bobfuru commented on modules/op-specifying-pvc-as-volumesource-in-workspaces.adoc:
```diff
@@ -0,0 +1,41 @@
+$ oc create -f <PersistentVolumeClaim-yaml-file-name.yaml>
+----
++
+Alternatively, you can also execute the YAML file directly from the Git repository:
```
s/you can also execute/you can execute/

Preeticp commented:
I have created a [Jira ](https://issues.redhat.com/browse/SRVKP-998) for the snippet changes and will send a follow up PR for it. IMHO that should not block this PR.

bobfuru reviewed (commented):
Overall, looks good with a few more minor revisions needed before merging. Makes sense about the code snippet issue and spec descriptions. Also, can you please squash commits? Thanks!

Preeticp commented:
Fixed comments and squashed @bobfuru Request to merge to 4.5 and 4.6

bobfuru commented:
LGTM!

bobfuru commented:
/cherrypick enterprise-4.6

bobfuru commented:
/cherrypick enterprise-4.5
